pkg/querier/frontend/frontend.go

```diff
@@ -345,18 +350,48 @@ func (f *Frontend) getNextRequest(ctx context.Context) (*request, error) {
 			continue
 		}
 
-		request := <-queue
+		/*
```

Comment:
I believe this entire new logic is over complicated for the use case. Moreover, returning `nil, nil` looks an anti-pattern to me: the `getNextRequest()` contract is that it "block if there are no requests" but with this PR changes there are conditions under which we do "return in some cases if there are no requests".

I'm wondering if the new logic could be implemented as:
1. Add a goto label before `for len(f.queues) == 0 && ctx.Err() == nil {`
2. Within the `for {`, if `request.originalCtx.Err() != nil` we do `continue`
3. We replace `panic("should never happen")` with `goto <label we've added at point 1>`

Makes sense?

Reply:
I like to goto idea -- we can maintain the same contract.

Reply:
@owen-d Do you have any update on this?

Reply:
@pracucci It should be updated. If you have small suggestions, feel free to add them, but I think this addresses the anti pattern you described earlier and aligns the function's contract to what it was before this pr.